Rework psa_generate_key

ronald-cron-arm opened the pull request:
## Description
Rework PSA key generation as described in the documentation file docs/architecture/psa-crypto-implementation-structure.md.

This work was primarily triggered by testing considerations. In terms of testing of the driver interface and its integration into the
PSA application interface implementation, with this PR:

- all key generation cryptographic operations are triggered by a call to the key generation driver interface. The call is the same for hardware and software implementations. The key generation driver interface is thus exercised by all the PSA unit tests that generate a key as opposed to just the driver wrapper tests before this PR.
- the all.sh test_psa_crypto_drivers component runs all PSA unit tests with all ECC/RSA key generation cryptographic operations handled by the transparent test driver.
- the all.sh test_psa_crypto_config_basic component runs all PSA unit tests with all ECC/RSA key import/export/generate cryptographic operations handled by the transparent test driver with the entry points to the ECC/RSA software implementations not included in the Mbed TLS library (the entry points are still in the library but accessed through test specific names).

## Status
READY

## Requires Backporting
NO, PSA only. 

## Steps to test or reproduce
CI testing is enough.

gilles-peskine-arm commented on tests/src/drivers/keygen.c:
```diff
@@ -57,58 +57,8 @@ psa_status_t test_transparent_generate_key(
-    else
-#endif /* MBEDTLS_ECP_C */
-    return( PSA_ERROR_NOT_SUPPORTED );
+    return( psa_generate_key_internal( attributes,
```
This introduces an extra dependency of the test code on the library code. Before, the test code only depended on the mbedtls API plus the mbedtls/psa glue function `mbedtls_to_psa_error`. Now the code depends on code in `psa_crypto.c`.

If Mbed TLS is compiled without software support for e.g. ECC through the PSA API, then under the [new PSA configuration mechanism](https://github.com/ARMmbed/mbedtls/blob/development/docs/proposed/psa-conditional-inclusion-c.md), the PSA code in the Mbed TLS library should not include support for ECC, which means that `psa_generate_key_internal` would not support ECC, and therefore this driver would not be usable for testing. But a configuration where Mbed TLS doesn't have the software implementation is precisely where it's the most important to use the test driver. So we would need to add yet another testing mode so that the library code is included in the build. But then we wouldn't really be testing the real product.

ronald-cron-arm replied:
Replying to this comment on the first version of the PR for the sake of completeness. The main focus of the first version of the PR was the test driver code and to avoid to rewrite in test drivers psa_crypto,c. The scope of the changes have changed since and their rationale explained in docs/architecture/psa-crypto-implementation-structure.md. Even if I think that this PR already improve significantly the testing of the `generate_key` entry point and the dispatch of key generation to a transparent driver (see the PR description), it still does not address the root issue (tracked by issue #3680) raised by this comment that the test driver `generate_key` entry points code is part of the MbedTLS crypto library.

gilles-peskine-arm replied:
This comment dates from before we settled on a test strategy that will involve throwing away much of the code in this file, right? Since this is short-term code, the extra dependency doesn't matter after all.

gilles-peskine-arm reviewed (changes requested):
This pull request changes a test driver to use more library code. I don't see how this can be compatible with the objective of the test driver. The goal of the test driver is to test how the library behaves when the software implementation is not included in the build and a driver is included instead. Ideally this test driver would not call any library code at all, but would instead have its own implementation of the algorithms that it supports. This PR is a move in the opposite direction.

I get it that you want to reduce code duplication. But the duplication here is between library code and test code, and these are not interchangeable. It is a requirement of the test architecture that the implementation of the cryptographic mechanisms in the test driver are included in the test code and not in the library.

One way to achieve that while not having separate cryptographic source code for the test drivers could be to compile a separate instance of the library for testing, and use a linker script to remain the functions to avoid conflicts. (I'm not saying this is the way to go, just saying it's a possible way.) But then the driver could call `psa_generate_key`: I don't see how this refactoring would help.

Unless there's something I'm missing, the objective of this PR is fundamentally contrary to the purpose of the test driver.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+
+The [PSA Cryptography API specification](https://armmbed.github.io/mbed-crypto/psa/#application-programming-interface) defines an interface to cryptographic operations which the Mbed TLS library provides a reference implementation for. The PSA Cryptography API specification is complemented by the PSA driver interface specification which defines an interface for cryptoprocessor drivers.
+
+This document describes the high level organization of the Mbed TLS PSA Cryptography API implementation which is tighly related to the PSA driver interface.
```
Fixed in #3933 as this is part of #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+
+## The Mbed TLS PSA Cryptography API implementation core
+
+The core implements all the APIs as defined in the PSA Cryptography API specification but do not perform on its own any cryptographic operation. The core relies on PSA drivers to actually
```
Fixed in #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+## The Mbed TLS PSA Cryptography API implementation core
+
+The core implements all the APIs as defined in the PSA Cryptography API specification but do not perform on its own any cryptographic operation. The core relies on PSA drivers to actually
+perform the cryptographic operations. The core is responsible of:
```
Fixed in #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+perform the cryptographic operations. The core is responsible of:
+
+* the key store.
+* checking PSA API arguments and translate them into valid arguments for the necessary calls to the PSA driver interface.
```
Fixed in #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+
+* the key store.
+* checking PSA API arguments and translate them into valid arguments for the necessary calls to the PSA driver interface.
+* dispatchng the cryptographic operations to the appropriate PSA drivers.
```
Fixed in #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
```diff
@@ -0,0 +1,50 @@
+        return( status );
+
+    /* Post driver interface call processing: validation of the values returned
+     * by the driver, finalization ot the values to return to the caller,
```
Fixed in #3933.

ronald-cron-arm commented on docs/architecture/psa-crypto-implementation-structure.md:
````diff
@@ -0,0 +1,50 @@
+     * clean-up in case of error ... */
+}
+```
+Obviously, for some PSA APIs the implementation is more complicated with several potential conditionnal calls to different driver entry points but most of PSA API code aimed to be organized along those lines. The implementation of the `psa_driver_wrapper_<entry_point>` function is generated by the build system based on the JSON driver description files of the various PSA drivers making up the Mbe ...
````
Fixed in #3933.

ronald-cron-arm commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -254,14 +253,15 @@ psa_status_t psa_driver_wrapper_verify_hash( psa_key_slot_t *slot,
     psa_key_type_t key_type = attributes->core.type;
     size_t key_bits = attributes->core.bits;
 
+    *key_buffer_size = 0;
```
Indeed, moved to the previous commit.

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -1651,111 +1148,121 @@ psa_status_t psa_get_key_slot_number(
+    }
+    else
+    {
+        /* This shouldn't happen in the reference implementation, but
```
@gilles-peskine-arm the tests from line 1173 to 1175 seem to include all possible keys but DH keys, not sure exporting DH keys make sense though. Thus, do we need to check for DH keys above or update this comment?

gilles-peskine-arm replied:
The comment is correct. Our implementation is intended to support exporting all key types, and we need that in unit test. Anything that is supported in `import` should be supported in `export`.

DH isn't supported here because it isn't supported in the library.

ronald-cron-arm commented:
Added back the "needs: reviewer" tag as I don't think @gilles-peskine-arm  is going to do a complete review of it.

chris-jones-arm reviewed (changes requested):
Overall I'm quite happy with this PR, the changes seem sensible and useful. Only a few small styling changes to fix and then I'm happy to approve.

As I am still quite new to PSA I would appreciate a quick design review from whoever merges this to ensure nothing was missed.

ronald-cron-arm commented:
@chris-jones-arm Thanks a lot for your review. I've addressed your comments please have another look. Unfortunately I had to rebase the PR on top of development because of a merge conflict. I hope this will not complicate too much the re-review.

Otherwise, I will ask to Manuel, Janos or Gilles to do the final checks and merge the PR. The PR modifies the code according to the docs/architecture/psa-crypto-implementation-structure.md document that has been reviewed by Manuel and Gilles thus we already have some guarantees regarding the design though.

chris-jones-arm reviewed (dismissed):
Great, thanks for the changes. I'm happy with this now.

@daverodgman can you re-review the changes?

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -6010,38 +6010,73 @@ static psa_status_t psa_read_rsa_exponent( const uint8_t *domain_parameters,
-static psa_status_t psa_generate_key_internal(
-    psa_key_slot_t *slot, size_t bits,
-    const uint8_t *domain_parameters, size_t domain_parameters_size )
+static psa_status_t psa_get_key_size(
```
This comment is still relevant to the last version. I addressed the documentation part: see [commit](https://github.com/ARMmbed/mbedtls/pull/3744/commits/f43e1d05a0b9ac83590b34fb08212ca7c7bfbd06). I didn't change the code to use PSA_EXPORT_KEY_OUTPUT_SIZE (previously PSA_KEY_EXPORT_MAX_SIZE) as this would result in a change of buffer size, for example for ECC keys: PSA_KEY_EXPORT_ECC_[KEY_PAIR|PUBLIC_KEY]_MAX_SIZE(key_bits) instead of PSA_BITS_TO_BYTES(key_bits) ... Not sure if I need to address this in this PR, let me know if you think so.

gilles-peskine-arm replied:
I don't understand when there would be a difference in size. For ECC key pairs, it's the bit-size rounded up to a whole number of bytes in both cases.

ronald-cron-arm replied:
Ah yes sorry, I didn't check the definition of PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE(key_bits).

ronald-cron-arm replied:
I split the additional validation needed for the key type and length for key generation and the key buffer size calculation. The key buffer size calculation is done just using  PSA_EXPORT_KEY_OUTPUT_SIZE now, see [1120cc0](https://github.com/ARMmbed/mbedtls/pull/3744/commits/1120cc08c396bf4c86ae96e73db86c8c25220936).

ronald-cron-arm commented:
@chris-jones-arm and @daverodgman I addressed a comment from Gilles I missed before with a new commit [1120cc0](https://github.com/ARMmbed/mbedtls/pull/3744/commits/1120cc08c396bf4c86ae96e73db86c8c25220936). Please re-review.

chris-jones-arm reviewed (dismissed):
Comments have been addressed and looks good.

daverodgman commented on library/psa_crypto.c:
```diff
@@ -5997,24 +5997,21 @@ psa_status_t mbedtls_psa_inject_entropy( const uint8_t *seed,
 #endif /* MBEDTLS_PSA_INJECT_ENTROPY */
 
-/** Get the key buffer size for the key material in export format
+/** Validate the key type and size for a key generation
```
"Validate the key type and size for key generation", or "Validate the key type and size for a key generation operation" are possibly slightly better.

ronald-cron-arm replied:
Thanks. As I have to rebase I will do the change.

ronald-cron-arm replied:
Done

daverodgman reviewed (approved):
LGTM (minor comment on wording)

ronald-cron-arm commented:
I've re-based due to one conflict in psa_crypto.c and one conflict in key_management.c. The conflicts were due to the usage in code moved by the PR of mbedtls_ecc_group_of_psa() whose signature was changed by the recently merged #4072 PR.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -238,62 +238,56 @@ psa_status_t psa_driver_wrapper_verify_hash( psa_key_slot_t *slot,
- * \param[out] expected_size    On success, a byte size large enough to contain
- *                              the declared key.
+ * \param[in] attributes  The key attributes.
+ * \param[out] key_buffer_size  Minimum buffer size to contain the key material
```
If this is going to work for key import as well, I suggest making this an in/out variable for when the 'bits' attribute is set to 0 (typically on import) and autodetection of the key size is needed.

Specifically: when the bits attribute is set to 0, require that the value of `*key_buffer_size` is pre-set to the length of the data input for import, such that this function can return a worst-case key buffer size, and the import_key function call can set the bits attribute right (since import_key itself is not able to resize the key buffer).

ronald-cron-arm replied:
We can adapt this function to support the key import case as well. As this PR is only about key generation I would prefer to do this in a separate PR though.

stevew817 replied:
Well, this PR changes `psa_driver_wrapper_get_key_buffer_size` to be available externally. Doesn't it then make sense to thing about why it's being made external, which parts of the codebase are going to use it, and what those different parts require in terms of functionality of this entrypoint?

ronald-cron-arm replied:
I understand you point but:
. in this PR the documentation of `psa_driver_wrapper_get_key_buffer_size` clearly states that it is dedicated to key generation thus I don't think there is a need (would probably be better though) to go beyond that.
. given all PSA PRs that we currently merge, as this PR is quite large, it conflicts rapidly with development. As it is in a state suitable for merge now, I prefer to finalize it and do the change on `psa_driver_wrapper_get_key_buffer_size` in another PR. I will work on the change when this PR is merged.

stevew817 replied:
Guess I'm looking forward to your next PR then.

gilles-peskine-arm commented on library/psa_crypto.c:
```diff
@@ -6000,167 +6001,105 @@ psa_status_t mbedtls_psa_inject_entropy( const uint8_t *seed,
+
+#if defined(PSA_WANT_KEY_TYPE_ECC_KEY_PAIR)
+    if( PSA_KEY_TYPE_IS_ECC( type ) && PSA_KEY_TYPE_IS_KEY_PAIR( type ) )
+    {
+    }
```
Please add a comment explaining why this block is empty. An empty block looks like a mistake and is easily misled in refactorings.

gilles-peskine-arm reviewed (dismissed):
Consider this a design approval, as a step towards implementing https://github.com/ARMmbed/mbedtls/issues/3680 by calling separate copies of functions. This is only a step towards #3680 so please don't close that issue yet.

I didn't do a detailed code review, I just glanced through the code. I did find one small maintainability issue.

ronald-cron-arm commented:
@daverodgman @chris-jones-arm I've re-based again. There was only one easy to solve conflict in library/CMakeLists.txt. Gilles has done the design review and I've addressed the issue he raised (see last additional commit). Could you please re-review quite quickly such as another conflict does not pop up in the meantime and we can merge the PR?

daverodgman commented on library/psa_crypto.c:
```diff
@@ -6142,15 +6084,42 @@ psa_status_t psa_generate_key( const psa_key_attributes_t *attributes,
+     * hold the generated key material. */
+    if( slot->key.data == NULL )
+    {
+        if ( PSA_KEY_LIFETIME_GET_LOCATION( attributes->core.lifetime ) ==
```
better to declare key_buffer_size here?

chris-jones-arm replied:
Although `key_buffer_size` is only used in scope of this if statement I don't see it as a problem that it is declared above with the other declarations. In fact I think it could even make it harder to read and find the declarations and thus I prefer it the way it is.

ronald-cron-arm replied:
I will think about this a bit more and may address it in a follow-up PR.

daverodgman commented on library/psa_crypto_rsa.c:
```diff
@@ -258,6 +259,66 @@ static psa_status_t rsa_export_public_key(
+        return( PSA_SUCCESS );
+    }
+
+    /* Mbed TLS encodes the public exponent as an int. For simplicity, only
```
"values that fit in a 32-bit integer, which is larger than int on just about every platform" - more accurate to say "... is at least as large as int ..."? Bit of a nit-pick but on first reading this confused me.

gilles-peskine-arm replied:
I wrote this sentence originally, it's just been moved here. `int` only fits 31-bit positive integers. The comment should have read something like “larger than what fits in int”.

ronald-cron-arm replied:
I will think about this a bit more and may address it in a follow-up PR.

chris-jones-arm reviewed (approved):
LGTM.
